puffin/src/lib.rs

```diff
@@ -489,34 +557,101 @@ impl GlobalProfiler {
         let current_frame_index = self.current_frame_index;
         self.current_frame_index += 1;
 
-        let new_frame =
-            match FrameData::new(current_frame_index, std::mem::take(&mut self.current_frame)) {
-                Ok(new_frame) => Arc::new(new_frame),
-                Err(Error::Empty) => {
-                    return; // don't warn about empty frames, just ignore them
-                }
-                Err(err) => {
-                    eprintln!("puffin ERROR: Bad frame: {err:?}");
-                    return;
+        // Scope details are registered here because we want them to be available for sinks to read during the new frame.
+        // The [`FrameData`] will get a delta of scopes that where reported this frame.
+        let mut scope_deltas =
+            Vec::with_capacity(self.new_scopes.len() + self.new_custom_scopes.len());
+
+        // Firstly add the new scopes registered through the macros.
+        for scope_detail in self.new_scopes.drain(..) {
+            scope_deltas.push(self.scope_collection.insert(Arc::new(scope_detail)));
+        }
+
+        // Secondly add the new scopes registered manually.
+        scope_deltas.extend(self.new_custom_scopes.drain(..));
+
+        let current_frame_scope = std::mem::take(&mut self.current_frame);
+
+        // Thirdly add a full snapshot of all scopes if requested.
+        // Its easier to send full delta to every sink.
+        let propagate_full_delta = std::mem::take(&mut self.propagate_all_scope_details);
+
+        if propagate_full_delta {
+            self.scope_collection.scopes_by_id(|scopes| {
+                for (_, details) in scopes {
+                    scope_deltas.push(details.clone());
                 }
-            };
+            });
+        }
+
+        let new_frame = match FrameData::new(
+            current_frame_index,
+            current_frame_scope,
+            scope_deltas,
+            propagate_full_delta,
+        ) {
+            Ok(new_frame) => Arc::new(new_frame),
+            Err(Error::Empty) => {
+                return; // don't warn about empty frames, just ignore them
+            }
+            Err(err) => {
+                eprintln!("puffin ERROR: Bad frame: {err:?}");
+                return;
+            }
+        };
 
         self.add_frame(new_frame);
     }
 
     /// Manually add frame data.
     pub fn add_frame(&mut self, new_frame: Arc<FrameData>) {
+        for delta in &new_frame.scope_delta {
+            self.scope_collection.insert(delta.clone());
+        }
+
         for sink in self.sinks.values() {
             sink(new_frame.clone());
         }
     }
 
+    /// Insert custom scopes into puffin.
+    /// Scopes details should only be registered once for each scope and need be inserted before being reported to puffin.
```

Comment:
This docstring should also document that the `scope_id`: s for the scopes are expected to be `None`